sdk/go/pulumi-language-go/main.go

```diff
@@ -157,10 +161,89 @@ func newLanguageHost(engineAddress, tracing string) pulumirpc.LanguageRuntimeSer
 	}
 }
 
+// modInfo is the useful portion of the output from `go list -m -json all`
+// with respect to plugin acquisition
+type modInfo struct {
+	Path    string
+	Version string
+}
+
+func (m *modInfo) getPlugin() (*pulumirpc.PluginDependency, error) {
+	if !strings.HasPrefix(m.Path, "github.com/pulumi/pulumi-") {
+		return nil, errors.New("module is not a pulumi provider")
+	}
+
+	// github.com/pulumi/pulumi-aws/sdk/... => aws
+	pluginPart := strings.Split(m.Path, "/")[2]
+	name := strings.SplitN(pluginPart, "-", 2)[1]
+
+	v, err := semver.ParseTolerant(m.Version)
+	if err != nil {
+		return nil, errors.New("module does not have semver compatible version")
+	}
+	// downgrade to zero for the patch
+	// pinning to a specific commit can result in a minor version that doesn't exist
+	// v1.29.1-0.20200403140640-efb5e2a48a86 (first commit after 1.29.0 release)
+	version := fmt.Sprintf("v%v.%v.%v", v.Major, v.Minor, 0)
```

Comment:
In the case that a precise version was specified - this seems wrong. 

Also in general - I’m not positive this won’t work - for Kubernetes for example I think dev builds are prerelease for major version bumps?  And is dependent on some very specific things about how we currently version dev tags.  Anything more reliable we could do here?  

Perhaps no and this is still the best we can do heuristically and is worth it?

Reply:
Good point. I'll lift some of the pseduo version parsing code out of internal to make this more robust.